### PR TITLE
Merge nominal parsing into the one function

### PR DESCRIPTION
**Pull request by bnbarham**
All nominals had extremely similar parse functions. Merge them into the one function so that improvements to one improves them all.

### CODE REVIEW

**Review by DougGregor — APPROVED** [review]
Thank you! This was bugging me

**Comment by bnbarham** [comment]
@swift-ci please test

File: Sources/SwiftParser/Nominals.swift [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    return parser.parseGenericParameters()
+  }
+}
+
+extension Parser {
+  /// Parse a nominal type decl declaration - class, struct, enum, or actor.
+  mutating func parseNominalTypeDeclaration<T>(
+    _ T: T.Type,
+    _ attrs: DeclAttributes,
+    _ handle: RecoveryConsumptionHandle
```
Comment (ahoppen) [inline Sources/SwiftParser/Nominals.swift:150]: Any reason why you decided to make these unlabeled? Also, while you’re at it, `handle` could probably be better named `introducerHandle` or `keywordHandle` (I have a mild preference for the former but don’t care too much).

**Review by ahoppen — APPROVED** [review]

File: Sources/SwiftParser/Nominals.swift (hunk 1 above)
Comment (bnbarham) [reply Sources/SwiftParser/Nominals.swift:150]: It was less a decision and more a copy paste of what we already had (with a rename):
```
 public mutating func parseClassDeclaration(
     _ attrs: DeclAttributes,
     _ handle: RecoveryConsumptionHandle
   ) -> RawClassDeclSyntax {
```

I don't have a strong preference here, how much do you care?

Comment (ahoppen) [reply Sources/SwiftParser/Nominals.swift:150]: If I wrote this, I would have labeled them. But if you want to merge this as-is, go ahead.

Comment (bnbarham) [reply Sources/SwiftParser/Nominals.swift]: I'm pretty sure you did add `handle` 🤔. Anyway, I went ahead and add them. Hopefully you're okay with `for` :P

File: Tests/SwiftParserTest/Declarations.swift [hunk 2]
```diff
@@ -804,7 +804,7 @@ final class DeclarationTests: XCTestCase {
 
   func testDontRecoverFromDeclKeyword() {
     AssertParse(
-      "func foo(first second 1️⃣third 3️⃣struct4️⃣: Int) {}",
+      "func foo(first second 1️⃣third 2️⃣struct3️⃣: Int4️⃣) {}",
```
Comment (bnbarham) [inline Tests/SwiftParserTest/Declarations.swift:807]: This was previously parsing just `struct` with all unexpected, but now (after also checking for start of line) parses the remained as a `struct` instead.

File: Sources/SwiftParser/Nominals.swift [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+extension Parser {
+  /// Parse a nominal type decl declaration - class, struct, enum, or actor.
+  mutating func parseNominalTypeDeclaration<T>(
+    for T: T.Type,
+    attrs: DeclAttributes,
+    introucerHandle: RecoveryConsumptionHandle
+  ) -> T where T: NominalTypeDeclarationTrait {
+    let (unexpectedBeforeIntroducerKeyword, introducerKeyword) = self.eat(introucerHandle)
+    let (unexpectedBeforeName, name) = self.expectIdentifier(keywordRecovery: true)
+    if unexpectedBeforeName == nil && name.isMissing && self.currentToken.isAtStartOfLine {
```
Comment (bnbarham) [inline Sources/SwiftParser/Nominals.swift:154]: And to be clear - `self.currentToken.isAtStartOfLine` is the change here.

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test

File: Tests/SwiftParserTest/translated/DeprecatedWhereTests.swift [hunk 4]
```diff
@@ -174,33 +174,33 @@ final class DeprecatedWhereTests: XCTestCase {
...
       diagnostics: [
         // TODO: Old parser expected error on line 1: 'where' clause next to generic parameters is obsolete, Fix-It replacements: 60 - 76 = '', 83 - 83 = ' where T: ProtoC'
         // TODO: Old parser expected error on line 1: 'protocol<...>' composition syntax has been removed
         DiagnosticSpec(locationMarker: "2️⃣", message: "expected inherited type in generic parameter"),
         DiagnosticSpec(locationMarker: "2️⃣", message: "expected '>' to end generic parameter clause"),
         DiagnosticSpec(locationMarker: "2️⃣", message: "expected parameter clause in function signature"),
-        DiagnosticSpec(locationMarker: "3️⃣", message: "expected identifier and member block in protocol"),
-        DiagnosticSpec(locationMarker: "3️⃣", message: "extraneous code '<ProtoA, ProtoB> where T: ProtoC>(x: T) {}' at top level"),
+        DiagnosticSpec(locationMarker: "3️⃣", message: "expected identifier in protocol"),
+        DiagnosticSpec(locationMarker: "4️⃣", message: "unexpected code '>(x: T)' in protocol"),
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/DeprecatedWhereTests.swift:186]: I figure this is no worse. But... also isn't great.

File: Tests/SwiftParserTest/translated/DollarIdentifierTests.swift [hunk 5]
```diff
@@ -57,7 +57,8 @@ final class DollarIdentifierTests: XCTestCase {
       """,
       diagnostics: [
         // TODO: Old parser expected error on line 2: '$' is not an identifier; use backticks to escape it, Fix-It replacements: 9 - 10 = '`$`'
-        DiagnosticSpec(message: "expected identifier and member block in class"),
+        DiagnosticSpec(message: "expected identifier in class"),
+        DiagnosticSpec(message: "unexpected code '$' in class"),
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/DollarIdentifierTests.swift:61]: I suppose we should add `dollarIdentifier` to `expectIdentifier`? I can do that in another PR. I probably shouldn't have added the start of line check in this one either.

Comment (ahoppen) [reply Tests/SwiftParserTest/translated/DollarIdentifierTests.swift:61]: I think/hope this should be fixed by https://github.com/apple/swift-syntax/pull/998.

Comment (bnbarham) [reply Tests/SwiftParserTest/translated/DollarIdentifierTests.swift:61]: Ah thank you. I'll rebase 🙇

**Comment by bnbarham** [comment]
@swift-ci please test

**Pull request merged**